odl/util/utility.py

```diff
@@ -833,15 +995,436 @@ def signature_string(posargs, optargs, sep=', ', mod=''):
+              np.array(value).real.astype('int64') != value and
+              modifier in ('', '!s', '!r')):
+            fmt = '{{:.{}}}'.format(precision)
+            value_str = fmt.format(value)
+            optargs_conv.append('{}={}'.format(name, value_str))
         else:
             fmt = '{{{}}}'.format(modifier)
             value_str = fmt.format(value)
             optargs_conv.append('{}={}'.format(name, value_str))
 
-    if optargs_conv:
-        parts.append(opt_sep.join(optargs_conv))
+    return tuple(posargs_conv), tuple(optargs_conv)
 
-    return part_sep.join(parts)
+
+def _separators(strings, linewidth):
+    """Return separators that keep joined strings within the line width."""
+    if len(strings) <= 1:
+        return ()
+
+    indent_len = 4
+    separators = []
+    cur_line_len = indent_len + len(strings[0]) + 1
+    if cur_line_len + 2 <= linewidth and '\n' not in strings[0]:
+        # Next string might fit on same line
+        separators.append(', ')
+        cur_line_len += 1  # for the extra space
+    else:
+        # Use linebreak if string contains newline or doesn't fit
+        separators.append(',\n')
+        cur_line_len = indent_len
+
+    for i, s in enumerate(strings[1:-1]):
+
+        cur_line_len += len(s) + 1
+
+        if '\n' in s:
+            # Use linebreak before and after if string contains newline
+            separators[i] = ',\n'
+            cur_line_len = indent_len
+            separators.append(',\n')
+
+        elif cur_line_len + 2 <= linewidth:
+            # This string fits, next one might also fit on same line
+            separators.append(', ')
+            cur_line_len += 1  # for the extra space
+
+        elif cur_line_len <= linewidth:
+            # This string fits, but next one won't
+            separators.append(',\n')
+            cur_line_len = indent_len
+
+        else:
+            # This string doesn't fit but has no newlines in it
+            separators[i] = ',\n'
+            cur_line_len = indent_len + len(s) + 1
+
+            # Need to determine again what should come next
+            if cur_line_len + 2 <= linewidth:
+                # Next string might fit on same line
+                separators.append(', ')
+            else:
+                separators.append(',\n')
+
+    cur_line_len += len(strings[-1])
+    if cur_line_len + 1 > linewidth or '\n' in strings[-1]:
+        # This string and a comma don't fit on this line
+        separators[-1] = ',\n'
+
+    return tuple(separators)
+
+
+def repr_string(outer_string, inner_strings, allow_mixed_seps=True):
+    r"""Return a pretty string for ``repr``.
+
+    The returned string is formatted such that it does not extend
+    beyond the line boundary if avoidable. The line width is taken from
+    NumPy's printing options that can be retrieved with
+    ``np.get_printoptions()``. They can be temporarily overridden
+    using the `npy_printoptions` context manager. See Examples for details.
```

Comment:
This does not create a link

Reply:
It should, since it's one of our functions in the same module